source/windows/bcrypt_aes.c

```diff
@@ -0,0 +1,1121 @@
+    key_data.dwMagic = BCRYPT_KEY_DATA_BLOB_MAGIC;
+    key_data.dwVersion = BCRYPT_KEY_DATA_BLOB_VERSION1;
+    key_data.cbKeyData = (ULONG)key->len;
+
+    struct aws_byte_buf key_data_buf;
+    aws_byte_buf_init(&key_data_buf, allocator, sizeof(key_data) + key->len);
+    aws_byte_buf_write(&key_data_buf, (const uint8_t *)&key_data, sizeof(key_data));
+    aws_byte_buf_write(&key_data_buf, key->buffer, key->len);
+
+    BCRYPT_KEY_HANDLE key_handle;
+    status = BCryptImportKey(
+        algHandle, NULL, BCRYPT_KEY_DATA_BLOB, &key_handle, NULL, 0, key_data_buf.buffer, (ULONG)key_data_buf.len, 0);
+
+    aws_byte_buf_clean_up_secure(&key_data_buf);
+
+    if (!NT_SUCCESS(status)) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
+    }
+
+    return key_handle;
+}
+
+static void s_aes_default_destroy(struct aws_symmetric_cipher *cipher) {
+    struct aes_bcrypt_cipher *cipher_impl = cipher->impl;
+
+    aws_byte_buf_clean_up_secure(&cipher->key);
+    aws_byte_buf_clean_up_secure(&cipher->iv);
+    aws_byte_buf_clean_up_secure(&cipher->tag);
+    aws_byte_buf_clean_up_secure(&cipher->aad);
+
+    /* clean_up_secure exists in versions of aws-c-common that don't check that the
+       buffer has a buffer and an allocator before freeing the memory. Instead,
+       check here. If it's set the buffer was owned and needs to be cleaned up, otherwise
+       it can just be dropped as it was an alias.*/
+    if (cipher_impl->working_iv.allocator) {
+        aws_byte_buf_clean_up_secure(&cipher_impl->working_iv);
+    }
+
+    aws_byte_buf_clean_up_secure(&cipher_impl->overflow);
+    aws_byte_buf_clean_up_secure(&cipher_impl->working_mac_buffer);
+
+    if (cipher_impl->key_handle) {
+        BCryptDestroyKey(cipher_impl->key_handle);
+        cipher_impl->key_handle = NULL;
+    }
+
+    if (cipher_impl->auth_info_ptr) {
+        aws_mem_release(cipher->allocator, cipher_impl->auth_info_ptr);
+        cipher_impl->auth_info_ptr = NULL;
+    }
+
+    aws_mem_release(cipher->allocator, cipher_impl);
+}
+
+/* just a utility function for setting up windows Ciphers and keys etc....
+   Handles copying key/iv etc... data to the right buffers and then setting them
+   on the windows handles used for the encryption operations. */
+static int s_initialize_cipher_materials(
+    struct aes_bcrypt_cipher *cipher,
+    const struct aws_byte_cursor *key,
+    const struct aws_byte_cursor *iv,
+    const struct aws_byte_cursor *tag,
+    const struct aws_byte_cursor *aad,
+    size_t iv_size,
+    bool is_ctr_mode,
+    bool is_gcm) {
+
+    if (!cipher->cipher.key.len) {
+        if (key) {
+            aws_byte_buf_init_copy_from_cursor(&cipher->cipher.key, cipher->cipher.allocator, *key);
+        } else {
+            aws_byte_buf_init(&cipher->cipher.key, cipher->cipher.allocator, AWS_AES_256_KEY_BYTE_LEN);
+            aws_symmetric_cipher_generate_key(AWS_AES_256_KEY_BYTE_LEN, &cipher->cipher.key);
```

Comment:
should we pull AWS_AES_256_KEY_BYTE_LEN into args or at least state in doc that this function assumes 256 byte keys. could save some pain in the future with someone using it for different key length